Lib/contextlib.py

```diff
@@ -14,6 +15,14 @@
            "chdir"]
 
 
+_PyErr_SetHandledException = ctypes.pythonapi.PyErr_SetHandledException
+_PyErr_SetHandledException.argtypes = [ctypes.py_object]
+def _set_current_exception(exc):
+    if exc is not None and not isinstance(exc, BaseException):
+        raise TypeError(f"{exc!r} is not an exception")
+    _PyErr_SetHandledException(exc)
```

Comment:
Is there any way to do this without ctypes?  Would it make sense to add a "private" (leading underscore) function to the sys module?

```suggestion
def _set_current_exception(exc):
    if exc is not None and not isinstance(exc, BaseException):
        raise TypeError(f"{exc!r} is not an exception")
    sys._restore_exception(exc)
```

That said, I don't have enough familiarity with our exception handling machinery to know if something like `sys._restore_exception()` (or `sys._set_exception()`) would invite abuses and end up an attractive nuisance.  @iritkatriel, what do you think?

Perhaps there is some other place it could go that would be less likely to attract abuse, like some new `_contextlib` C extension?  (I'd want feedback from @ncoghlan before we added a `_contextlib` module.)

Reply:
Do also note that, as much as possible, we want stdlib .py modules to work in other Python implementations, like pypy.  Relying on ctypes makes that problematic generally, which is why basically no stdlib .py modules rely on ctypes.  At the very least this code should have a non-ctypes fallback.  I'm just not sure what your alternative is since you'd probably already be using that instead of ctypes.  (Falling back to the existing confusing behavior isn't really a desireable solution.)

Reply:
I'm definitely not familiar with the nuances of how best to use the C API in a cross-runtime way. I went with `ctypes` because it was simple to implement without having to make changes to the C API.

I've just pushed a commit that implements `sys._set_exception()` (and some tests) and uses it instead of using `ctypes`. Open to other suggestions as well.

Reply:
Considering this resolved by implementing `sys._set_exception()`. @iritkatriel / @ericsnowcurrently: Feel free to unresolve if you would prefer an alternate approach.